abci/strategies/aggregator/vote_aggregator.go

```diff
@@ -0,0 +1,214 @@
+		logger: logger,
+		priceAggregator: aggregator.NewDataAggregator(
+			aggregator.WithAggregateFnFromContext(aggregateFn),
+		),
+		currencyPairStrategy: strategy,
+	}
+}
+
+type DefaultVoteAggregator struct {
+	// validator address -> currency-pair -> price
+	priceAggregator *aggregator.DataAggregator[string, map[oracletypes.CurrencyPair]*big.Int]
+
+	// decoding prices / currency-pair ids
+	currencyPairStrategy currencypair.CurrencyPairStrategy
+
+	logger log.Logger
+}
+
+func (dva *DefaultVoteAggregator) AggregateOracleVotes(ctx sdk.Context, votes []Vote) (map[oracletypes.CurrencyPair]*big.Int, error) {
+	// Reset the price aggregator and set the aggregationFn to use the latest application-state.
+	dva.priceAggregator.ResetProviderData()
+
+	// Iterate through all vote extensions and consolidate all price info before
+	// aggregating.
+	for _, vote := range votes {
+		consAddrStr := vote.ConsAddress.String()
+
+		if err := dva.addVoteToAggregator(ctx, consAddrStr, vote.OracleVoteExtension); err != nil {
+			dva.logger.Error(
+				"failed to add vote to aggregator",
+				"validator_address", consAddrStr,
+				"err", err,
+			)
+
+			return nil, err
+		}
+	}
+
+	// Compute the final prices for each currency pair.
+	dva.priceAggregator.AggregateDataFromContext(ctx)
+	prices := dva.priceAggregator.GetAggregatedData()
+
+	dva.logger.Info(
+		"aggregated oracle data",
+		"num_prices", len(prices),
+	)
+
+	return prices, nil
+}
+
+// addVoteToAggregator consolidates the oracle data from a single validator
+// into the price aggregator. The oracle data is provided in the form of a vote
+// extension. The vote extension contains the prices for each currency pair that
+// the validator is providing for the current block.
+func (dva *DefaultVoteAggregator) addVoteToAggregator(ctx sdk.Context, address string, oracleData vetypes.OracleVoteExtension) error {
+	if len(oracleData.Prices) == 0 {
+		return nil
+	}
+
+	// Format all of the prices into a map of currency pair -> price.
+	prices := make(map[oracletypes.CurrencyPair]*big.Int, len(oracleData.Prices))
+	for cpID, priceBz := range oracleData.Prices {
+		if len(priceBz) > slinkyabci.MaximumPriceSize {
+			return fmt.Errorf("price bytes are too long: %d", len(priceBz))
+		}
+
+		// Convert the asset into a currency pair.
+		cp, err := dva.currencyPairStrategy.FromID(ctx, cpID)
+		if err != nil {
+			dva.logger.Debug(
+				"failed to convert currency pair id to currency pair",
+				"currency_pair_id", cpID,
+				"err", err,
+			)
+
+			// If the currency pair is not supported, continue.
+			//
+			// TODO: Should we return an error here instead?
+			continue
+		}
+
+		price, err := dva.currencyPairStrategy.GetDecodedPrice(ctx, cp, priceBz)
+		if err != nil {
+			dva.logger.Debug(
+				"failed to decode price",
+				"currency_pair_id", cpID,
+				"err", err,
+			)
+
+			// If the price cannot be decoded, continue.
+			//
+			// TODO: Should we return an error here instead?
+			continue
```

Comment:
is this a carryover or a new TODO?

Reply:
ya this is a carry-over

Reply:
Would you prefer that we remove the TODO or change the behavior?

Reply:
Linear: https://linear.app/skip/issue/BLO-865/resolve-potential-errors-in-preblock-when-aggregating-votes-prices